### ISSUE TITLE
Wrong version of `COMCTL32.DLL` found.

### ISSUE THREAD

**Issue by MarekKnapek** [hfiref0x/WinDepends#7]
**Describe the bug**
There are multiple versions of `COMCTL32.DLL` on my machine. Version 5.xxx and version 6.xxx, which one gets loaded at run-time depends on SxS / Activation Context / Manifest.

**To Reproduce**
 - Start the `WinDepends.exe` tool.
 - Analyze the `c:\Windows\System32\comdlg32.dll` file. This is the 64bit file.
 - Notice that it depends on `COMCTL32.DLL` file.
 - But according to your tool it depends on the 6.xxx version located at `C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.20348.2849_none_947683b1eef1cae2\COMCTL32.dll`.
 - But according to my tool it depends on the 5.xxx version located at `C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.20348.2849_none_377a7937165aefa3\comctl32.dll`.

**Expected behavior**
The tool should display correct dependency.

**Screenshots**
no

**Desktop (please complete the following information):**
 - Windows Server 2022 x64.
 - Microsoft Windows [Version 10.0.20348.2849]

**Files**
no

**Additional context**
Sadly I have no time to develop my tool further. https://github.com/MarekKnapek/DependencyViewer I spent quite some time on exactly this kind of issue, so I'm pretty confident I got it correctly (if not correctly, then at least more correct than your tool). Feel free to borrow / steal source code and / or any ideas from my project.

**Comment by hfiref0x** [comment]
Should be fixed in https://github.com/hfiref0x/WinDepends/commit/ae1f67a487195d69c47051ad0faea5be552201cf. If not, feel free to mention it again.

The main problem here was that I wasn't sure how WinForms manages visual styles, and because of that, we forced comctl 6.0 entry in the program manifest. This, as you understand, makes it context default. It turns out WinForms can handle visual styles via additional context almost like your program does with your independent comctl context.

**Comment by MarekKnapek** [comment]
Both tools still differ.

Analyze the `C:\Windows\System32\comdlg32.dll` file. (amd64 arch)

My tool says it depends (among other things) on the `C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.20348.2849_none_377a7937165aefa3\comctl32.dll` file.
Your tool says it depends (among other things) on the: `C:\Windows\System32\COMCTL32.dll` file.

I believe my tool is right and your tool is wrong. My tools finds it thanks to code located in `dependency_locator.cpp` file, lines 33 and 43.

**Comment by hfiref0x** [comment]
That is debatable what is correct display here.

COMDLG32.DLL has no CREATEPROCESS_MANIFEST_RESOURCE_ID embedded, but instead has some manifest 123 with comctl 6.0 as a dependency. While the stated valid range of manifest IDs is between 1 and 16.

Thus, creating an activation context (with CREATEPROCESS_MANIFEST_RESOURCE_ID or ISOLATIONAWARE_MANIFEST_RESOURCE_ID) on that DLL will fail. Such a failure causes the current code in WinDepends to skip resolving at the PathFromWinSXS routine and go to the next variants, which succeed later in the system32 directory lookup.

Technically, COMCTL32.DLL from the system32 directory is a hard link to \amd64_microsoft-windows-shell-comctl32-v5_31bf3856ad364e35_10.0.19041.4355_none_ae563eca8d31ba11\comctl32.dll (as of my test environment), which is version 5.82 of comctl32.

Now take the original Dependency Walker and analyze the same comdlg32.dll. It will attempt to create activation context for it looking for ISOLATIONAWARE_MANIFEST_RESOURCE_ID (since this is dll - it checks the file extension), and fail on that with INVALID_HANDLE_VALUE as result. Next it will go to SearchPath and show you version 5.82 with the source path in WinSXS because SearchPath takes into account the current activation context, and depends.exe does not have a manifest (which brings default activation context to work). Is this correct? From the target program perspective - yes. 

Now, if we add a manifest with comctl 6.0 to Dependency Walker and reanalyze comdlg32.dll, it will show you c:\windows\winsxs\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.4355_none_60b8b9eb71f62e16\COMCTL32.DLL as a dependency as it now has an activation context with this DLL. Is this correct? From the target program perspective - yes.

Ever wondered why depends.exe never had any manifests despite being updated until Windows Vista? Now you know why.

Since the target file is a DLL and no suitable manifest is found, I believe the correct behavior in case of an activation context failure on that file is to skip any side-by-side API query functions and go to other methods, which will eventually bring you to the same 5.82 comctl version but as a hard link in system32.

**Comment by MarekKnapek** [comment]
> hard link

Does not matter. Since it is different name, it could have different content. And on older Windows you could not even enumerate a file "hard link siblings" without guessing "sibling" name first.

> ... causes the current code in WinDepends to skip ...

Is that really correct behavior? I'm not sure. One of us has it wrong, back then when I wrote my code, I tested it thoroughly and I was pretty satisfied. Today I'm not sure. All I want is that "somebody, oh my god, I don't care who, has it correctly". Few years back I was hoping for https://github.com/lucasg/Dependencies to deliver superb application but he also ran out of steam. Now I'm hoping for you. That's why I'm creating these bug reports.

> ... depends.exe does not have a manifest (which brings default activation context to work). Is this correct? ...

I believe yes.

> ... add a manifest with comctl 6.0 to Dependency Walker and reanalyze comdlg32.dll, it will show you COMCTL32.DLL v6. Is this correct? ...

I believe yes.

> ... ever wondered why depends.exe never had any manifests ...

No, I believe it must have none in order to work correctly. 

> What to do in such case?

I suggest to study ntdll.dll and loader snaps. But I don't have enough energy for it right now.

**Comment by hfiref0x** [comment]
> I suggest to study ntdll.dll and loader snaps.

What I'm talking here is: for side-by-side module it will depend on currently active activation context.
DLL itself doesn't exist as separate loader object. If you load it in process it will be resolved according to that process environment settings such as activation context. 

If I create empty application with no manifest and force it load comdlg32.dll here it will load as dependency 5.82 comctl32. If I add manifest with 6.0 version it will load instead it. Both are correct.

In our case we cannot create context from that dll, because there is no allowed side-by-side information in it (no 1 or 2 manifests id in it). How does it will be resolved here depends now only on current process activation context stack. If it default with no 6.0 version specified it will resolved as 5.82, otherwise 6.0, both are technically correct.

From the user perspective you have dll file that is resolving according to viewer program side-by-side settings. Does it correct? I don't think so.

**Comment by MarekKnapek** [comment]
> ... DLL itself doesn't exist as separate loader object ...

OK, I didn't expressed myself precisely. What I meant is that your tool reports `comctl32.dll` dependency incorrectly. No matter if you analyze a `*.dll` file or an `*.exe` file. Now I will clarify my bug report.

Steps to reproduce:

 - Copy & paste source code of a simple 64bit MASM application. [1]
 - Compile the application by invoking these two commands.
   - `ml64 tiny.asm /c`
   - `link tiny.obj user32.lib comctl32.lib /ENTRY:main`
 - Analyze `tiny.exe` by your tool.
 - Your tool reports dependency on `C:\Windows\System32\COMCTL32.dll` file.
 - Execute the `tiny.exe` program, it will not exit immediately, thus you have chance to examine running process by other tools, such as by [System Informer](https://github.com/winsiderss/systeminformer).
 - Realize that `c:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.20348.2849_none_377a7937165aefa3\comctl32.dll` file is being loaded.

[1]
```
extern __imp_InitCommonControls:qword
InitCommonControls equ <__imp_InitCommonControls>

extern __imp_MessageBoxA:qword
MessageBoxA equ <__imp_MessageBoxA>


.data

caption db "info",0
text db "Press OK to exit application.",0


.code

main proc frame
	push rbp
	sub  rsp, 10h
	mov  rbp, rsp
.endprolog

	sub  rsp, 60h

	call InitCommonControls

	xor   r9d, r9d
	lea   r8,  caption
	lea   rdx, text
	xor   rcx, rcx
	call  MessageBoxA

	add rsp, 60h
	add rsp, 10h
	pop rbp
	ret
main endp

end
```
